AppCenter/AppCenter/Internals/Context/AuthToken/MSAuthTokenContext.m

```diff
@@ -53,62 +52,168 @@ + (void)resetSharedInstance {
+    [authTokenHistory addObject:newAuthToken];
+
+    // Cap array size at max available size const (deleting from beginning).
+    if ([authTokenHistory count] > kMSMaxAuthTokenArraySize) {
+      [authTokenHistory removeObjectsInRange:(NSRange){0, [authTokenHistory count] - kMSMaxAuthTokenArraySize}];
+      MSLogWarning([MSAppCenter logTag], @"Size of the token history is exceeded. The oldest token has been removed.");
+    }
+    isNewAccount = ![self.accountId isEqual:accountId];
 
     // Don't invoke the delegate while locking; it might be locking too and deadlock ourselves.
     synchronizedDelegates = [self.delegates allObjects];
+
+    // Save new array.
+    [self setAuthTokenHistory:authTokenHistory];
   }
   for (id<MSAuthTokenContextDelegate> delegate in synchronizedDelegates) {
     if ([delegate respondsToSelector:@selector(authTokenContext:didSetNewAuthToken:)]) {
       [delegate authTokenContext:self didSetNewAuthToken:authToken];
     }
-    if (isNewUser && [delegate respondsToSelector:@selector(authTokenContext:didSetNewAccountIdWithAuthToken:)]) {
+    if (isNewAccount && [delegate respondsToSelector:@selector(authTokenContext:didSetNewAccountIdWithAuthToken:)]) {
       [delegate authTokenContext:self didSetNewAccountIdWithAuthToken:authToken];
     }
   }
 }
 
-- (BOOL)clearAuthToken {
-  NSArray *synchronizedDelegates;
-  BOOL clearedExistingUser = NO;
+- (void)addDelegate:(id<MSAuthTokenContextDelegate>)delegate {
   @synchronized(self) {
-    if (!self.authToken) {
-      return NO;
-    } else if (self.homeAccountId) {
-      clearedExistingUser = YES;
-    }
-    self.authToken = nil;
-    self.homeAccountId = nil;
+    [self.delegates addObject:delegate];
+  }
+}
 
-    // Don't invoke the delegate while locking; it might be locking too and deadlock ourselves.
-    synchronizedDelegates = [self.delegates allObjects];
+- (void)removeDelegate:(id<MSAuthTokenContextDelegate>)delegate {
+  @synchronized(self) {
+    [self.delegates removeObject:delegate];
   }
-  for (id<MSAuthTokenContextDelegate> delegate in synchronizedDelegates) {
-    if ([delegate respondsToSelector:@selector(authTokenContext:didSetNewAuthToken:)]) {
-      [delegate authTokenContext:self didSetNewAuthToken:nil];
-    }
-    if (clearedExistingUser && [delegate respondsToSelector:@selector(authTokenContext:didSetNewAccountIdWithAuthToken:)]) {
-      [delegate authTokenContext:self didSetNewAccountIdWithAuthToken:nil];
+}
+
+- (nullable NSString *)authToken {
+  NSArray<MSAuthTokenInfo *> *authTokenHistory = [self authTokenHistory];
+  MSAuthTokenInfo *latestAuthTokenInfo = authTokenHistory.lastObject;
+  return latestAuthTokenInfo.authToken;
+}
+
+- (nullable NSString *)accountId {
+  NSArray<MSAuthTokenInfo *> *authTokenHistory = [self authTokenHistory];
+  MSAuthTokenInfo *latestAuthTokenInfo = authTokenHistory.lastObject;
+  return latestAuthTokenInfo.accountId;
+}
+
+- (NSMutableArray<MSAuthTokenValidityInfo *> *)authTokenValidityArray {
```

Comment:
It feels like this method would need some comments.

Reply:
Its out there already, in a header file:
```Objective-C
/*
 * Returns array of auth tokens validity info.
 *
 * @return Array of MSAuthTokenValidityInfo
 */
- (NSMutableArray<MSAuthTokenValidityInfo *> *)authTokenValidityArray;
```